In [138]:
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.graph_objects as go

In [139]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [140]:
%sql postgresql://postgres:postgres@localhost:5430/eskom_db

In [141]:
%%sql 
DROP TABLE IF EXISTS eskom_data;
CREATE TABLE  eskom_data(
    date_time_hour_beginning DATE,
    original_res_forecast_before_lockdown NUMERIC,
    residual_forecast  NUMERIC,	
    rsa_contracted_forecast	 NUMERIC,
    dispatchable_generation	NUMERIC ,
    residual_demand NUMERIC,
    rsa_contracted_demand  NUMERIC,
    international_exports  NUMERIC,
    international_imports  NUMERIC,
    thermal_generation  NUMERIC,
    nuclear_generation  NUMERIC,
    eskom_gas_generation  NUMERIC,	
    eskom_ocgt_generation  NUMERIC,	
    hydro_water_generation  NUMERIC,
    pumped_water_generation	 NUMERIC,
    ils_usage  NUMERIC,
    manual_load_reduction NUMERIC,
    ios_excl_ils_and_mlr  NUMERIC,
    dispatchable_ipp_ocgt  NUMERIC,
    eskom_gas_sco  NUMERIC,
    eskom_ocgt_sco  NUMERIC,
    hydro_water_sco	 NUMERIC,
    pumped_water_sco  NUMERIC,
    pumping	 NUMERIC,
    wind  NUMERIC,
    pv  NUMERIC,
    csp  NUMERIC,
    other_re  NUMERIC,	
    total_re  NUMERIC,	
    wind_installed_capacity	 NUMERIC,
    pv_installed_capacity  NUMERIC,	
    csp_installed_capacity  NUMERIC,
    other_re_installed_capacity	 NUMERIC,
    total_re_installed_capacity	 NUMERIC,
    installed_eskom_capacity  NUMERIC,
    total_pclf  NUMERIC,
    total_uclf  NUMERIC,
    total_oclf  NUMERIC,
    total_uclf_and_oclf	 NUMERIC,
    non_comm_sentout  NUMERIC,	
    drakensberg_gen_unit_hours  NUMERIC,
    palmiet_gen_unit_hours  NUMERIC,	
    ingula_gen_unit_hours NUMERIC
    
);


 * postgresql://postgres:***@localhost:5430/eskom_db
Done.
Done.


[]

In [142]:
%%sql 
COPY eskom_data(date_time_hour_beginning,original_res_forecast_before_lockdown,residual_forecast,rsa_contracted_forecast,dispatchable_generation,residual_demand,rsa_contracted_demand,international_exports,international_imports,thermal_generation,nuclear_generation,eskom_gas_generation,eskom_ocgt_generation,hydro_water_generation,pumped_water_generation,ils_usage,manual_load_reduction,ios_excl_ils_and_mlr,dispatchable_ipp_ocgt,eskom_gas_sco,eskom_ocgt_sco,hydro_water_sco,pumped_water_sco,pumping,wind,pv,csp,other_re,total_re,wind_installed_capacity,pv_installed_capacity,csp_installed_capacity,other_re_installed_capacity,total_re_installed_capacity,installed_eskom_capacity,total_pclf,total_uclf,total_oclf,total_uclf_and_oclf,non_comm_sentout,drakensberg_gen_unit_hours,palmiet_gen_unit_hours,ingula_gen_unit_hours)
FROM '/Users/ds_learner16/Documents/Eskom/Eskom_Data_Analysis/ESK2033.csv'
DELIMITER ','
CSV HEADER
;

 * postgresql://postgres:***@localhost:5430/eskom_db
37704 rows affected.


[]

In [143]:
%%sql 
DROP TABLE IF EXISTS generation,demand,capacity;
CREATE TABLE demand
AS 
SELECT
    EXTRACT(YEAR FROM date_time_hour_beginning) as year,
    EXTRACT(MONTH FROM date_time_hour_beginning) as month,
    EXTRACT(DAY FROM date_time_hour_beginning) as day,
    date_time_hour_beginning,
    residual_demand,
    rsa_contracted_demand
FROM eskom_data;
CREATE TABLE generation
AS 
SELECT
    EXTRACT(YEAR FROM date_time_hour_beginning) as year,
    EXTRACT(MONTH FROM date_time_hour_beginning) as month,
    EXTRACT(DAY FROM date_time_hour_beginning) as day,
    dispatchable_generation,
    international_exports,
    international_imports,
    thermal_generation,
    nuclear_generation,
    eskom_gas_generation,	
    eskom_ocgt_generation,	
    hydro_water_generation,
    pumped_water_generation,
     dispatchable_ipp_ocgt,
    pumping,
    wind,
    pv,
    csp,
    other_re,	
    total_re,	
    drakensberg_gen_unit_hours,
    palmiet_gen_unit_hours,	
    ingula_gen_unit_hours
FROM eskom_data;
CREATE TABLE capacity
AS 
SELECT
    EXTRACT(YEAR FROM date_time_hour_beginning) as year,
    EXTRACT(MONTH FROM date_time_hour_beginning) as month,
    EXTRACT(DAY FROM date_time_hour_beginning) as day,
    wind_installed_capacity,
    pv_installed_capacity,	
    csp_installed_capacity,
    other_re_installed_capacity,
    total_re_installed_capacity,
    installed_eskom_capacity
FROM eskom_data;




 * postgresql://postgres:***@localhost:5430/eskom_db
Done.
37704 rows affected.
37704 rows affected.
37704 rows affected.


[]

In [144]:
%%sql
DROP TABLE IF EXISTS Loadshedding;
CREATE TABLE loadshedding (
    created_at TIMESTAMP,
    stage INT
      
);

 * postgresql://postgres:***@localhost:5430/eskom_db
Done.
Done.


[]

In [145]:
%%sql
COPY Loadshedding(created_at,stage)
FROM '/Users/ds_learner16/Documents/Eskom/Eskom_Data_Analysis/EskomSePush_history.csv'
DELIMITER ','
CSV HEADER;

 * postgresql://postgres:***@localhost:5430/eskom_db
670 rows affected.


[]

In [164]:
%%sql 
DROP TABLE IF EXISTS load_shedding_clean;
CREATE TABLE load_shedding_clean
AS
SELECT 
    to_char(created_at,'Mon') as month_name,
    to_char(created_at,'Day') as day_name,
    EXTRACT(YEAR FROM created_at) as year,
    EXTRACT(MONTH FROM created_at) as month,
    EXTRACT(DAY FROM created_at) as day,
    EXTRACT(HOUR FROM created_at) as hour,
    stage

FROM Loadshedding;



 * postgresql://postgres:***@localhost:5430/eskom_db
Done.
670 rows affected.


[]

# Number of Rows and Columns 
count rows = 42824 estimated before clean after 37704
removed forecast dates added average for original res before lockdown

count cols = 41

## Which year was the demand high

In [147]:
high_demand = %sql SELECT year as "year" , AVG(residual_demand) as "residual_demand", AVG(rsa_contracted_demand) as "rsa_contracted_demand" FROM demand GROUP BY year ORDER BY year
high_demand

 * postgresql://postgres:***@localhost:5430/eskom_db
5 rows affected.


[(Decimal('2018'), Decimal('25767.131784696970'), Decimal('26998.603180303030')),
 (Decimal('2019'), Decimal('25219.673938127854'), Decimal('26542.384553082192')),
 (Decimal('2020'), Decimal('23696.602886042805'), Decimal('25117.220372153916')),
 (Decimal('2021'), Decimal('24196.092115525114'), Decimal('25932.202556392694')),
 (Decimal('2022'), Decimal('24394.402998125000'), Decimal('26094.779150833333'))]

In [148]:
fig = px.line(high_demand,x="year",y=["residual_demand","rsa_contracted_demand"],title="Residual vs RSA Contracted Demand")
fig.update_layout(
    xaxis= dict(
        dtick = 1
    )
)
fig.show()

## Which of the renewable energy generation methods have the highest capacity

In [149]:
renew_capacity = %sql SELECT year as "year",AVG(wind_installed_capacity) as "wind",AVG(pv_installed_capacity) as "pv",AVG(csp_installed_capacity) as "csp",AVG(other_re_installed_capacity) as "other" FROM eskom_clean GROUP BY year ORDER BY year
renew_capacity

 * postgresql://postgres:***@localhost:5430/eskom_db
5 rows affected.


[(Decimal('2018'), Decimal('1474.1900000000000000'), Decimal('311.6363636363636364'), Decimal('21.7800000000000000'), Decimal('3886.6987272727272727')),
 (Decimal('2019'), Decimal('1474.1900000000000000'), Decimal('491.7808219178082192'), Decimal('21.7800000000000000'), Decimal('4067.5108219178082192')),
 (Decimal('2020'), Decimal('1780.3017486338797814'), Decimal('500.0000000000000000'), Decimal('21.7800000000000000'), Decimal('4439.6280874316939891')),
 (Decimal('2021'), Decimal('2194.0434246575342466'), Decimal('500.0000000000000000'), Decimal('25.0073972602739726'), Decimal('5380.9399726027397260')),
 (Decimal('2022'), Decimal('2212.0900000000000000'), Decimal('500.0000000000000000'), Decimal('41.8300000000000000'), Decimal('5893.4680000000000000'))]

In [150]:
fig1 = px.line(renew_capacity,x="year",y=["wind","pv","csp","other"],title="Average of renewable energy capacity")
fig1.update_layout(
    xaxis= dict(
        dtick = 1
    )
)
fig1.show()

## Which Year , month , hour had the most loadshedding 

In [151]:
yearly = %sql SELECT year as "year" ,count(case when stage = 1 then 1 else null end) as "stage_1",count(case when stage = 2 then 1 else null end) as "stage_2",count(case when stage = 3 then 1 else null end) as "stage_3",count(case when stage = 4 then 1 else null end) as "stage_4",count(case when stage = 5 then 1 else null end) as "stage_5" FROM load_shedding_clean GROUP BY year ORDER by year
yearly
fig2 = px.line(yearly,x="year",y=["stage_1","stage_2","stage_3","stage_4","stage_5"])
fig2.show()

 * postgresql://postgres:***@localhost:5430/eskom_db
6 rows affected.


In [152]:
monthly = %sql SELECT month,month_name as "month_name" ,count(case when stage = 1 then 1 else null end) as "stage_1",count(case when stage = 2 then 1 else null end) as "stage_2",count(case when stage = 3 then 1 else null end) as "stage_3",count(case when stage = 4 then 1 else null end) as "stage_4",count(case when stage = 5 then 1 else null end) as "stage_5" FROM load_shedding_clean GROUP BY month,month_name ORDER by month
monthly
fig2 = px.line(monthly,x="month_name",y=["stage_1","stage_2","stage_3","stage_4","stage_5"])
fig2.show()

 * postgresql://postgres:***@localhost:5430/eskom_db
12 rows affected.


In [177]:
dayly = %sql SELECT day_name as "day_name" ,count(case when stage = 1 then 1 else null end) as "stage_1",count(case when stage = 2 then 1 else null end) as "stage_2",count(case when stage = 3 then 1 else null end) as "stage_3",count(case when stage = 4 then 1 else null end) as "stage_4",count(case when stage = 5 then 1 else null end) as "stage_5" FROM load_shedding_clean GROUP BY day_name ORDER by day_name
dayly
fig2 = px.line(dayly,x="day_name",y=["stage_1","stage_2","stage_3","stage_4","stage_5"])
fig2.show()

 * postgresql://postgres:***@localhost:5430/eskom_db
7 rows affected.


In [154]:
hourly = %sql SELECT hour as "hour" ,count(case when stage = 1 then 1 else null end) as "stage_1",count(case when stage = 2 then 1 else null end) as "stage_2",count(case when stage = 3 then 1 else null end) as "stage_3",count(case when stage = 4 then 1 else null end) as "stage_4",count(case when stage = 5 then 1 else null end) as "stage_5" FROM load_shedding_clean GROUP BY hour ORDER by hour
hourly
fig2 = px.line(hourly,x="hour",y=["stage_1","stage_2","stage_3","stage_4","stage_5"])
fig2.show()

 * postgresql://postgres:***@localhost:5430/eskom_db
21 rows affected.


## How much energy is generated yearly by eskom and other enties in the country

In [155]:
%%sql
select * from generation limit 5

 * postgresql://postgres:***@localhost:5430/eskom_db
5 rows affected.


[(Decimal('2018'), Decimal('4'), Decimal('1'), Decimal('20237'), Decimal('1215.902'), Decimal('1120'), Decimal('19444'), Decimal('931'), Decimal('0'), Decimal('0'), Decimal('360'), Decimal('0'), Decimal('0'), Decimal('454.964'), Decimal('0'), Decimal('17.555'), Decimal('12.539'), Decimal('485.058'), Decimal('2077.72'), Decimal('81.8'), Decimal('36.9'), Decimal('30.41')),
 (Decimal('2018'), Decimal('4'), Decimal('1'), Decimal('19744'), Decimal('1203.474'), Decimal('1106'), Decimal('19297'), Decimal('930'), Decimal('0'), Decimal('0'), Decimal('360'), Decimal('0'), Decimal('0'), Decimal('392.958'), Decimal('0'), Decimal('39.131'), Decimal('12.404'), Decimal('444.493'), Decimal('2077.72'), Decimal('83'), Decimal('38.5'), Decimal('32.85')),
 (Decimal('2018'), Decimal('4'), Decimal('1'), Decimal('19631'), Decimal('1177.571'), Decimal('1117'), Decimal('19165'), Decimal('931'), Decimal('0'), Decimal('0'), Decimal('360'), Decimal('0'), Decimal('0'), Decimal('337.358'), Decimal('0'), Decimal('39.163'), Decimal('12.082'), Decimal('388.603'), Decimal('2077.72'), Decimal('83.8'), Decimal('40.3'), Decimal('35.6')),
 (Decimal('2018'), Decimal('4'), Decimal('1'), Decimal('19731'), Decimal('1184.312'), Decimal('1118'), Decimal('19279'), Decimal('930'), Decimal('0'), Decimal('0'), Decimal('360'), Decimal('0'), Decimal('0'), Decimal('297.234'), Decimal('0'), Decimal('38.942'), Decimal('12.278'), Decimal('348.454'), Decimal('2077.72'), Decimal('85'), Decimal('42'), Decimal('37.76')),
 (Decimal('2018'), Decimal('4'), Decimal('1'), Decimal('19890'), Decimal('1197.271'), Decimal('1108'), Decimal('19369'), Decimal('930'), Decimal('0'), Decimal('0'), Decimal('360'), Decimal('0'), Decimal('0'), Decimal('309.978'), Decimal('0'), Decimal('25.024'), Decimal('12.488'), Decimal('347.49'), Decimal('2077.72'), Decimal('85.8'), Decimal('43.7'), Decimal('40.32'))]

## Do we import more energy or export more energy?

## Which one of the energy generation methods yeild the most power in GW

## Thermal power produces the most power but is showing decline in recent years

## Which is the highest performing power station out of the 3 stations